### ISSUE TITLE
Keras/TensorFlow

### ISSUE THREAD

**Issue by susantkumarsahoo** [susantkumarsahoo/StudentPerformancePrediction-End-To-End#6]
# Keras/TensorFlow Complete Reference Guide

A comprehensive reference guide for all major techniques, layers, and algorithms available in Keras/TensorFlow.

---

# Neural Network Layers

## ActivityRegularization - Applies L1/L2 regularization on output

---

# Convolutional Layers

## ZeroPadding3D - Adds zeros around data (3D)

---

# Pooling Layers

## GlobalAveragePooling3D - Global average pooling (3D)

---

# Recurrent Layers

## TimeDistributed - Applies layer to every temporal slice

---

# Normalization Layers

## UnitNormalization - Normalizes to unit norm

---

# Regularization Layers

## AlphaDropout - Dropout for SELU activation

---

# Attention Layers

## MultiHeadAttention - Multi-head attention mechanism

---

# Merging Layers

## Dot - Computes dot product

---

# Embedding Layers

## TextVectorization - Converts text to integer sequences

---

# Activation Functions

## Mish - x * tanh(softplus(x))

---

# Optimizers

## Ftrl - Follow The Regularized Leader

---

# Loss Functions for Regression

## LogCosh - Log of hyperbolic cosine

---

# Loss Functions for Classification

## CosineSimilarity - Cosine similarity loss

---

# Accuracy Metrics

## TopKCategoricalAccuracy - Top-K accuracy

---

# Regression Metrics

## CosineSimilarity - Cosine similarity

---

# Classification Metrics

## FalseNegatives - Count of false negatives

---

# Probabilistic Metrics

## KLDivergence - KL divergence metric

---

# Callbacks

## BackupAndRestore - Backs up model for fault tolerance

---

# Regularizers

## L1L2 - Combined L1 and L2 regularization

---

# Initializers

## LecunUniform - LeCun uniform initialization

---

# Image Preprocessing

## RandomBrightness - Random brightness adjustment

---

# Text Preprocessing

## IntegerLookup - Maps integers to integers

---

# Numerical Preprocessing

## Hashing - Hashing trick for features

---

# Model Types

## Model Subclassing - Custom models via inheritance

---

# Pre-trained Models: Image Classification

## ConvNeXt - Modern ConvNet design

---

# Pre-trained Models: Object Detection

## EfficientDet - Scalable object detection

---

# Pre-trained Models: NLP

## GPT - Generative pre-trained transformer (via TF Hub)

---

# Data Augmentation

## Rescaling - Scales pixel values

---

# Training Utilities

## set_weights() - Sets layer weights

---

# Data Loading

## tf.data.Dataset - TensorFlow dataset pipeline

---

# Mixed Precision Training

## LossScaleOptimizer - Scales gradients for mixed precision

---

# Distributed Training

## ParameterServerStrategy - Asynchronous training

---

# Custom Components

## Custom Training Loops - Use GradientTape for control

---

## Installation

```bash
pip install tensorflow
# or
pip install keras
```

## Quick Example

```python
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

# Build model
model = keras.Sequential([
    layers.Dense(128, activation='relu', input_shape=(784,)),
    layers.Dropout(0.2),
    layers.Dense(10, activation='softmax')
])

# Compile
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

# Train
model.fit(x_train, y_train, epochs=5, batch_size=32)

# Evaluate
model.evaluate(x_test, y_test)
```

---

## Resources

- [[Keras Official Documentation](https://keras.io/)](https://keras.io/)
- [[TensorFlow Documentation](https://www.tensorflow.org/)](https://www.tensorflow.org/)
- [[Keras API Reference](https://keras.io/api/)](https://keras.io/api/)
- [[TensorFlow Tutorials](https://www.tensorflow.org/tutorials)](https://www.tensorflow.org/tutorials)

---

**Last Updated:** October 2025
